capstone/src/test/java/com/google/sps/RecommenderAlgorithmTest.java

```diff
@@ -0,0 +1,149 @@
+package com.google.sps;
+ 
+import static com.google.appengine.api.datastore.FetchOptions.Builder.withLimit;
+ 
+import org.junit.Assert;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.sps.data.Room;
+import com.google.sps.data.User;
+import com.google.sps.data.Tag;
+import com.google.sps.data.RecommenderAlgorithm;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.sps.service.DatabaseService;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.HashMap;
+ 
+@RunWith(JUnit4.class)
+public final class RecommenderAlgorithmTest {
+    
+    private final LocalServiceTestHelper helper =
+        new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+    private static final List<String> TEST_CATEGORY_TAGS = Tag.CATEGORY_TAGS;
+
+    private static User TEST_USER_SMITH;
+    private static User TEST_USER_JOHN;
+    private static User TEST_USER_BILL;
+    private static User TEST_USER_SAM;
+    private static Room TEST_ROOM_POETRY;
+    private static Room TEST_ROOM_RUN;
+    private static Room TEST_ROOM_COOK; 
+    private static Room TEST_ROOM_COOK_TWO;
+
+    @Before
+    public void setUp() {
+        helper.setUp();
+        TEST_USER_SMITH = new User("1234", "Smith");
+        TEST_USER_JOHN = new User("2341", "John");
+        TEST_USER_BILL = new User("3412", "Bill");
+        TEST_USER_SAM = new User("4123", "Sam");
+        DatabaseService.save(TEST_USER_SMITH.getUserEntity());
+        DatabaseService.save(TEST_USER_JOHN.getUserEntity());
+        DatabaseService.save(TEST_USER_BILL.getUserEntity());
+        DatabaseService.save(TEST_USER_SAM.getUserEntity());
```

Comment:
Unit tests should not interact with the database at all. (In order to stay "unit"). They should be as small in scope as possible and only test a single method. If knowing that something was "saved" to the database is a critical part of what needs to be tested, a mock object should be used.

Sometimes writing such a mock object and all of the setup and expectations that requires is complicated and annoying, which can be informative about how you design your methods in the first place. Each method should be easy to test, so if it can avoid talking to the database at all, that would be ideal. (You can have separate out the logic part of the method and test that, then have an outer function that calls that along with doing whatever database saving needs to be done)

Reply:
Got it. My original thinking was that since these methods pulls entities from Datastore, I should mimic that behavior. It makes more sense to use mock objects though. Thanks for the tip!